File: javademo/src/main/java/zlc/season/javademo/CustomApplication.java
package zlc.season.javademo;

import android.app.Application;
import android.content.Intent;
import android.os.Handler;
import android.util.Log;
import android.widget.Toast;

import java.net.ConnectException;
import java.net.SocketTimeoutException;

import io.reactivex.Observable;
import io.reactivex.Observer;
import io.reactivex.disposables.Disposable;
import io.reactivex.functions.BiFunction;
import io.reactivex.plugins.RxJavaPlugins;

public class CustomApplication extends Application {
    private static final String TAG = "CustomApplication";
    private Handler mainHandler;

    @Override
    public void onCreate() {
        super.onCreate();
        mainHandler = new Handler(getMainLooper());

        RxJavaPlugins.setOnObservableSubscribe(new BiFunction<Observable, Observer, Observer>() {
            @Override
            public Observer apply(Observable observable, Observer observer) throws Exception {
                return new ObservableSubscribeHooker(observer);
            }
        });

    }

    class ObservableSubscribeHooker<T> implements Observer<T> {
        private Observer<T> actual;

        public ObservableSubscribeHooker(Observer<T> actual) {
            this.actual = actual;
        }

        @Override
        public void onSubscribe(Disposable d) {
            actual.onSubscribe(d);
        }

        @Override
        public void onNext(T t) {
            hookOnNext(t);
            actual.onNext(t);
        }

        private void hookOnNext(T t) {
            if (t instanceof BaseResponse) {
                BaseResponse baseResponse = (BaseResponse) t;
                if (baseResponse.getCode() == 100) {

                    mainHandler.post(new Runnable() {
                        @Override
                        public void run() {
                            Toast.makeText(getApplicationContext(), "登录过期", Toast.LENGTH_SHORT).show();
                        }
                    });

                    Intent intent = new Intent(getApplicationContext(), LoginActivity.class);
                    intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
                    startActivity(intent);

                    throw new Exceptions.TokenExpired();
                }
            }
        }


        @Override
        public void onError(Throwable e) {

            if (e instanceof ConnectException) {
                Log.e(TAG, "Connect failed: ", e);
                mainHandler.post(new Runnable() {
                    @Override
                    public void run() {
                        Toast.makeText(getApplicationContext(), "无网络连接", Toast.LENGTH_SHORT).show();
                    }
                });
                actual.onError(new Exceptions.Offline());
                return;
            }

            if (e instanceof SocketTimeoutException) {
                Log.e(TAG, "Time out ", e);
                mainHandler.post(new Runnable() {
                    @Override
                    public void run() {
                        Toast.makeText(getApplicationContext(), "连接超时", Toast.LENGTH_SHORT).show();
                    }
                });
                actual.onError(new Exceptions.TimeOut());
                return;
            }

            //其余的异常处理...

            actual.onError(e);
        }

        @Override
        public void onComplete() {
            actual.onComplete();
        }
    }
}

File: kotlindemo/src/main/java/zlc/season/kotlindemo/CustomApplication.kt
package zlc.season.kotlindemo

import android.app.Application
import android.content.Intent
import android.os.Handler
import android.util.Log
import android.widget.Toast
import io.reactivex.Observer
import io.reactivex.disposables.Disposable
import io.reactivex.plugins.RxJavaPlugins
import java.net.ConnectException
import java.net.SocketTimeoutException

class CustomApplication : Application() {
    companion object {
        const val TAG = "CustomApplication"
    }

    var mainHandler: Handler? = null

    override fun onCreate() {
        super.onCreate()
        mainHandler = Handler(mainLooper)

        RxJavaPlugins.setOnObservableSubscribe { _, t2 ->
            ObservableSubscribeHooker(t2)
        }
    }

    inner class ObservableSubscribeHooker<T>(private var actual: Observer<T>) : Observer<T> {
        override fun onComplete() {
            actual.onComplete()
        }

        override fun onSubscribe(d: Disposable) {
            actual.onSubscribe(d)
        }

        override fun onNext(t: T) {
            hookOnNext(t)
            actual.onNext(t)
        }

        private fun hookOnNext(t: T) {
            if (t is BaseResponse) {
                if (t.code == 100) {

                    mainHandler!!.post {
                        Toast.makeText(applicationContext, "登录过期", Toast.LENGTH_SHORT).show()
                    }

                    val intent = Intent(applicationContext, LoginActivity::class.java)
                    intent.flags = Intent.FLAG_ACTIVITY_NEW_TASK
                    startActivity(intent)

                    throw TokenExpiredException()
                }
            }
        }

        override fun onError(e: Throwable) {
            if (e is ConnectException) {
                Log.e(TAG, "Connect failed: ", e)
                mainHandler!!.post({ Toast.makeText(applicationContext, "无网络连接", Toast.LENGTH_SHORT).show() })
                actual.onError(Exceptions.Offline())
                return
            }

            if (e is SocketTimeoutException) {
                Log.e(TAG, "Time out ", e)
                mainHandler!!.post({ Toast.makeText(applicationContext, "连接超时", Toast.LENGTH_SHORT).show() })
                actual.onError(Exceptions.TimeOut())
                return
            }

            //其余的异常处理...
            actual.onError(e)
        }
    }

    class TokenExpiredException : RuntimeException()
}